### PR TITLE
Refactor ros2 frame with tests

### PR DESCRIPTION
**Pull request by arturkamieniecki**
I've added an editor component for the ROS2 frame component. 

I've modified the implementation of the ROS2 frame component to use EBus to communicate between frames.

I've also added simple tests for the component that checks if the frameID and joint names are adequately constructed with namespaces.

### CODE REVIEW

File: Gems/ROS2/Code/CMakeLists.txt [hunk 1]
```diff
@@ -197,6 +197,36 @@ if(PAL_TRAIT_BUILD_TESTS_SUPPORTED)
         ly_add_googletest(
             NAME Gem::${gem_name}.Tests
         )
+
+        
+        # integration test for URDF importer
```
Comment (adamdbrw) [inline Gems/ROS2/Code/CMakeLists.txt:202]:
```suggestion
        # integration test for ROS2 Frame
```

**Review by adamdbrw — COMMENTED** [review]
Any adjustments needed in Robot Importer?

File: Gems/ROS2/Code/CMakeLists.txt [hunk 2]
```diff
@@ -197,6 +197,36 @@ if(PAL_TRAIT_BUILD_TESTS_SUPPORTED)
...
+                    AZ::AzTestShared
+                    AZ::AzToolsFramework
+                    Legacy::CryCommon
+                    Legacy::EditorCommon
+                    Legacy::Editor.Headers
+                    AZ::AzManipulatorTestFramework.Static
+                    Gem::ROS2.Static
+                    Gem::ROS2.Editor.Static
+                RUNTIME_DEPENDENCIES
+                    Gem::PhysX.Editor
```
Comment (michalpelka) [inline Gems/ROS2/Code/CMakeLists.txt:223]: I believe it is not needed.

**Review by michalpelka — COMMENTED** [review]
I see a problem with this change - the old, now game-only ROS2FrameComponent is still present in all prefabs. You need to remove/update in this PR.

File: Gems/ROS2/Code/Include/ROS2/Frame/ROS2FrameComponent.h [hunk 3]
```diff
@@ -50,55 +55,47 @@ namespace ROS2
...
+         //! @note Supplementary metadata for Joint components, necessary in some cases for joints addressed by name in ROS 2
+         //! @return The namespaced joint name, ready to send in a ROS2 message
         AZ::Name GetJointName() const;
 
-        //! Set the joint name
-        //! @note May be populated during URDF import or set by the user in the Editor view 
-        //! @param jointNameString does not include the namespace. The namespace prefix is added automatically.
+         //! Set the joint name
+         //! @note May be populated during URDF import or set by the user in the Editor view
+         //! @param jointNameString does not include th namespace. The namespace prefix is added automatically.
```
Comment (pijaro) [inline Gems/ROS2/Code/Include/ROS2/Frame/ROS2FrameComponent.h:65]:
```suggestion
         //! @param jointNameString does not include the namespace. The namespace prefix is added automatically.
```

File: Gems/ROS2/Code/Include/ROS2/Frame/ROS2FrameController.h [hunk 4]
```diff
@@ -0,0 +1,120 @@
...
+
+        const AZ::EntityId GetParentROS2FrameComponentId() const;
+
+        //! Return the frame id of this frame's parent. It can be useful to determine ROS 2 transformations.
+        //! @return Parent frame ID.
+        //! @note This also works with top-level frames, returning a global frame name.
+        //! @see GetGlobalFrameName().
+        AZStd::string GetParentFrameID() const;
+
+        void PopulateNamespace(bool isRoot, const AZStd::string& entityName);
```
[the remark this answers is not in the document]
Comment (arturkamieniecki) [reply Gems/ROS2/Code/Include/ROS2/Frame/ROS2FrameController.h:93]: This function cannot be editor only. If it was editor only dynamically spawning a prefab containing ROS2FrameComponents wouldn't properly setup its namespaces.

**Comment by arturkamieniecki** [comment]
I've chaned all of the current uses of the ROS2FrameComponent to the EBus implementation, removing boilerplate code.

I've also added the effective namespace for the ROS2FrameEditorComponent.

**Comment by arturkamieniecki** [comment]
I've also created a python script to update all prefabs to the new ROS2FrameEditorComponent.

```import sys
import json
import os

def find_and_replace(data):

    def search_for_components(item, foundComponents):
        if isinstance(item, dict):
            if "Components" in item:
                foundComponents = True
            if "ROS2FrameComponent" in item and foundComponents:
                modifiedElement = item["ROS2FrameComponent"]
                modifiedElement["$type"] = "ROS2FrameEditorComponent"
                del modifiedElement["m_template"]["$type"]
                configuration = {"Configuration" : modifiedElement["m_template"]}
                modifiedElement["Controller"] = configuration
                del modifiedElement["m_template"]
                del item["ROS2FrameComponent"]
                item["ROS2FrameEditorComponent"] = modifiedElement
            if "op" in item and item["op"] == "replace" and "path" in item:
                item["path"] = item["path"].replace("ROS2FrameComponent/m_template", "ROS2FrameEditorComponent/Controller/Configuration")
            for value in item.values():
                search_for_components(value, foundComponents)
        elif isinstance(item, list):
            for element in item:
                search_for_components(element, foundComponents)

    search_for_components(data, False)


def write_json_to_file(data, filename):
    try:
        with open(filename, 'w') as file:
            json.dump(data, file, indent=4)
            print(f"Modified data written to '{filename}'.")
    except Exception as e:
        print(f"Error writing to '{filename}': {e}")

def find_prefab_files(directory):
    prefab_files = []
    if os.path.isfile(directory) and directory.endswith(".prefab"):
        prefab_files.append(directory)
    elif os.path.isdir(directory):
        for root, dirs, files in os.walk(directory):
            for file in files:
                if file.endswith(".prefab"):
                    prefab_files.append(os.path.join(root, file))
    return prefab_files

def process_prefab_file(file_path):
    try:
        with open(file_path, 'r') as file:
            data = json.load(file)
            find_and_replace(data)
        with open(file_path, 'w') as file:
            json.dump(data, file, indent=4)
            print(f"Modified data written to '{file_path}'.")
    except Exception as e:
        print(f"Error processing '{file_path}': {e}")

if __name__ == "__main__":
    if len(sys.argv) != 2:
        print("Usage: python script.py <directory_or_file>")
    else:
        input_path = sys.argv[1]
        prefab_files = find_prefab_files(input_path)
        if not prefab_files:
            print(f"No .prefab files found in '{input_path}'.")
        else:
            for file_path in prefab_files:
                print(f"Processing '{file_path}'...")
                process_prefab_file(file_path)
```

To use this script simply provide it with the argument pointing to the .prefab file or the directory in which all .prefab files should be changed.

**Comment by arturkamieniecki** [comment]
Closed in favor of #593.

**Pull request closed without merging**